Major refactor, added 2 custom libraries, fixed overlay movement

0xdavidel opened the pull request:
This is probably too many changes for a single pull request, got carried away, whoops 
Here is the gist:
1. Created BetterStashTabAPI - This is a custom library to access POE stash tabs and parse them. 
2. Created BetterFilterAPI - This is a custom library to read and modify .filter files, it only knows how to add a custom section to the file and how to modify those custom sections. This allows for an easy way to modify any filter file and add the chaos recipe rules to it
3. Changed the chaos recipe comments to allow use with the BetterFilterAPI
4. Fixed the Overlay movement, now you can click and drag using the leftmost button
5. Added a Sync thread to synchronize between the local and remote stash states
6. Major code refactor and commenting the whole code - Basically, I made it more readable and maintainable
7. Extracted most "magic values" out of the code - To allow maintenance and easy changes magic values should not exits
8. Added a better filter selection dialog, now instead of selecting a folder (that is static btw), you choose what filter you want to change
9. Removed redundant code, part of the refactor, there were some unused functions
10. Changed the highlight code to not use exec()
10. Highlights will now use the actual item size and not a preset number - possible using the BetterStashTabAPI
11. Better debug printing by wrapping the original pprint function - no more need to check if we are in debugging and then print, I wrapped it
12. Better screen resolution config handling - now you can have spaces and use a big X if you want
13. Remade the overlay content to better show how many identified and unidentified items there are - a simple text swap and using the BetterStashTabAPI to check the Identified and Unidentified files
14. Changed the paths to work with the new folder structure

0xdavidel committed:
VSCode Format

0xdavidel committed:
Added vscode launch.json

0xdavidel committed:
Added .gitignore and added the log file to it

0xdavidel committed:
Added utils folder

0xdavidel committed:
Added BetterStashTabAPI.py

0xdavidel committed:
Added BetterFilterAPI.py

0xdavidel committed:
Added .pyc to .gitignore

0xdavidel committed:
Moved the different resources into separate folders

No need to crowd the base directory, Code at this point doesn't support it, don't use this commit

0xdavidel committed:
Changed way to much in a single commit

1. Using BetterStashTabAPI
2. Using BetterFilterAPI
3. Fixed the Overaly movement
4. Added a Sync thread
5. Major code refactor and commenting the whole code
6. Extracted most "magic values" out of the code
7. Added a better filter selection dialog
8. Removed redundant code
9. Changed the highlight code to not use exec()
10. Highlights will now use the actual item size and not a preset number
11. Better debug printing by wrapping the original pprint function
12. Better screen resolution config handling
13. Remade the overlay content to better show how many identified and unidentified items there are
14. Changed the paths to work with the new folder structure

0xdavidel committed:
Added a requirements.txt file

Now you can use pip install -r requirements to setup all the libraries needed

notablackbear commented on utils/BetterFilterAPI.py:
```diff
@@ -0,0 +1,219 @@
+
+    Get the filter folder path
+    This path is always the filter directory
+    Proof - https://www.pathofexile.com/forum/view-thread/1835293/page/1#p14200591
```
This post must be outdated. I've worked with multiple players who's filter paths are slightly different from expected due to non-english players. For example, a path would include `Dokuments` instead of the British English `Documents`.

My quick solution for when this occurred would be to have the user interactively choose the correct folder via a GUI dialog.

0xdavidel replied:
Not a problem to implement, Added to my TODO list.
10 minutes later - fixed.

0xdavidel commented:
A bug I found - Accidentaly I am using os.exit instead of using sys.exit.

this results in an error instead of a clean exit, but it doesn't break the functionality of the code, fixing it now

0xdavidel committed:
Fixed os.exit bug

0xdavidel committed:
Absolute path fix, no longer forcing old POE filter path

This should allow the ability to put full paths in the setup.ini file in case of custom filter directories

0xdavidel committed:
Fixed identified / unidentified retrival, changed the API of retrieve_all_by_tag a little

notablackbear commented:
So i'm trying to test this branch out, but i'm having a ton of trouble closing the program since the extra thread is not being handled on close. I am trying to figure out how to handle it, but you might be able to quicker.

0xdavidel commented:
Thats odd, the thread is set to daemon mode, one sec, ill reimplement the function so that it doesn't sleep for 10 seconds

0xdavidel committed:
Reimplemented sync thread to poll every 0.5 seconds

0xdavidel commented:
Try it now, it now sleeps only half a seconds and each time it checks if enough time elapsed from the previous stash_sync check

Nevermind, I see what you mean, the gui closes but the python keeps on running, ill see what I can do

0xdavidel committed:
Added a way to close the sync_thread / added an on_close function

0xdavidel commented:
Added a way to close it, not the best code but the whole GUI side of the app still needs a refactor so that the tkApp only handles the GUI and separate out the logic into a different class and functions that are not inside the GUI class.

notablackbear commented:
So i've been testing this out, and it works really well for the most part. I think it might be good to keep perfecting it in another branch before merging into master though? Here's a couple of notes:
- [ ] Overlay text is too large
- [ ] Doesn't save filter to Setup file after first time user selects it
- [ ] Location of Chaos Recipe filter blocks overrides other filter rules (here is example of influenced item with chaos recipe highlighting)
![BadHighlightingDueToFilterOrdering](https://user-images.githubusercontent.com/71364995/94740861-289ec100-0328-11eb-8793-7b2d2c1cd6bd.png)


I think a couple of these are easy to fix, like the text.

notablackbear commented on POE_QOL2.py:
```diff
@@ -327,393 +525,242 @@ def show_chaos(self):
+            self.chaos_filter_parsed=load_rules_from_base_filter(
+                chaos_filter_path)
+        except Exception as e:
+            debug_print("Exception reading chaos filter: {}".format(str(e)))
```
should be `self.debug_print(`
